### DISCUSSION TITLE
cpp/basics/static

### DISCUSSION THREAD

**Discussion by giscus** [imarvinle/CSGuide#65] [Announcements]
# cpp/basics/static

编程指北: 计算机系统学习指南, 操作系统, 计算机网络, C++, Java, 算法, 数据结构

https://csguide.cn/cpp/basics/static.html

**Comment by huangjiawei128** [comment]
其实可以简单理解成static是对函数/变量作用域的限制。

1. 函数内的 static 变量
生存期：整个程序的执行期间。
作用域：定义它的函数内部。
初始化时机：第一次执行到定义它的函数时进行初始化。

2. 类内的 static 成员变量
生存期：整个程序的执行期间。
作用域：类内部。
初始化时机：在第一次使用之前初始化（必须在类外部定义和初始化）。

3. 文件作用域的 static 变量（内部链接）
生存期：整个程序的执行期间。
作用域：定义它的源文件内部。
初始化时机：程序开始时进行初始化。

**Reply by Atri2333** [reply]
类内的 static 成员变量初始化时机和全局变量以及文件作用域的 static 变量是一样的，都是在程序开始（main函数执行前）初始化，即便不使用也会初始化。
